DAT-344 from https://github.com/GreaterLondonAuthority/Data_…

callum-oakley opened the pull request:
…for_London/pull/44

Created from https://github.com/GreaterLondonAuthority/Data_for_London/pull/44 by running

```
git format-patch --relative=src/ckanext-gla origin/develop --stdout > dat-344.mbox
```

in the main repository and then applying the patch to this repository with

```
git am ../../../Data_for_London-alt/dat-344.mbox
```

Commit:
DAT-344 - Move JS for accordions out of script tags and into a webasset

Understanding how to set up JS webassets was a first step needed before adding more JS to log search query results

madelinekosse committed:
DAT-344 - Log clicks on search results

When a user clicks on a search result, log the timestamp, search
parameters, sort order, and the index of the selected result in a file.

At present, this information will populate a file called search_logs.csv
in the container's app directory but we will probably want to change
that. This example shows the log file structure:

|time                      |query |sort                              |org|tags      |format|licence|package-id                                    |index|
|--------------------------|------|----------------------------------|---|----------|------|-------|----------------------------------------------|-----|
|2023-05-30 13:10:32.915699|test  |score desc, metadata_modified desc|   |          |      |       |covid-19-online-diary                         |21   |
|2023-05-30 13:12:20.372644|survey|views_recent desc                 |ons|employment|CSV   |ogl-v2 |hours-worked                                  |4    |

Note that only clicks on search results are logged, so selecting an item
from the default dataset page, even if it is filtered for organisation
or sorted by a custom field, does not add a line to the file.

madelinekosse committed:
DAT-344 - Add accordion.js

Commit 708d1c0dbdf9427d2b71156129777196575ea5a6 moved the accordion javascript
into its own webasset, however the file was mistakenly omitted from the commit
so I have added it here.

Commit:
DAT-344 - Update accordion.js to use CKAN module system

madelinekosse committed:
DAT-344 - Start search result indices at 1

The previous implementation considered the first result index 0, when it should
be given as 1.

madelinekosse committed:
DAT-350 Add API endpoint to download log file

Adds an endpoint at `/api/action/search_logs` that serves the log file
containing user searches and selected datasets. This endpoint is
available only to admin users.

Because all action endpoint responses are wrapped in a standard response
format automatically, the response will not be a csv format file attachment
- it is nested within the `result` key of a json response.

madelinekosse committed:
DAT-351 - Download search logs from admin page

Adds a link, in the sidebar of the administrator page, to download the
search logs as a CSV file. Ideally, there would be an additional tab in
the administrator console for this, but this proved complicated so I
went for the simplest possible way to get the functionality in there for
now.

kjoshi commented on ckanext/gla/templates/snippets/package_list.html:
```diff
@@ -14,21 +14,29 @@
             {% endif %}
             {% set nonfavourites = h.remove_favourites(g.userobj, request, packages) %}
-                <ul class="{{ list_class or 'dataset-list list-unstyled' }}">
+                <ul class="{{ list_class or 'dataset-list list-unstyled' }}" data-module="gla_search_result_logging">
```
I'm not sure the `data-module` is doing anything here.`search_result_logging.js` doesn't start with: `ckan.module('gla_search_result_logging',...)` so I don't think a module with that name gets registered with CKAN.

I don't think the module is really necessary anyway if you're just calling `logResultClick` directly in `package_item`
```suggestion
                <ul class="{{ list_class or 'dataset-list list-unstyled' }}">
```

kjoshi commented on ckanext/gla/templates/admin/index.html:
```diff
@@ -0,0 +1,36 @@
+    </div>
+    <div class="module-content">
+        <p>System administrators can download a log of all user searches, and the results that they selected
+            as a csv file <a id="downloadlink"  data-module="gla_search_result_log_download">here</a>.</p>
```
Similar comment to `package_item.html` - I don't think the `data-module` is doing anything here.
```suggestion
            as a csv file <a id="downloadlink">here</a>.</p>
```

kjoshi commented on ckanext/gla/search.py:
```diff
@@ -34,3 +41,36 @@ def debug(context, data_dict={}):
+
+def _result_index(page, index_in_page):
+    page_idx = 0 if _empty_or_none(page) else int(page) - 1
+    return page_idx * 20 + int(index_in_page) + 1
```
The number of datasets per page can be set in the [config](https://github.com/ckan/ckan/blob/master/ckan/config/config_declaration.yaml#L1078), so it might be best to link this to that config value in case it get's changed at some point?

Also, according to the docs `loop.index` (in `package_list.html`) starts at 1 already:
![image](https://github.com/GreaterLondonAuthority/dfl-ckanext/assets/1217533/52dff114-7fe4-4dca-b2fc-4b350648f60c)
so I'm pretty sure we don't need the extra `+1` here

```suggestion
    from ckan.common import config
    return page_idx * config.get('ckan.datasets_per_page') + int(index_in_page)
```

kjoshi reviewed (changes requested):
I think I’ve found a bit of a nicer way of handling the file download.
I’ve created a new branch with some [example code](https://github.com/GreaterLondonAuthority/dfl-ckanext/commit/cfcd571bdfab6a1fe49694875a21178f56a759c2).

TLDR: We need a CKAN `view` to serve the file (by using `flask.send_file`), rather than an `action`.
My example code is very rough but hopefully you can see the idea. 
It would mean that the `get_logs` action could be replaced by the new view function and the `<script>` in `admin/index.html` could just be replaced with a link to the new route.

If you think this seems like a good idea maybe you could incorporate it into this PR? 

Another idea I had was to make the log file path configurable using [IConfigDeclaration](https://docs.ckan.org/en/2.10/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IConfigDeclaration) ([example here](https://docs.ckan.org/en/2.10/maintaining/configuration.html#declaring-config-options)) so that it's easier to get access to that variable in different files, and maybe easier to change in future. (E.g. we could create a new `ckanext.gla.search_logs_path` config variable)

But that could be overkill if this isn't a permanent solution...

madelinekosse committed:
DAT-334 - Use view instead of actions API to fetch files

See [Kiran's review](https://github.com/GreaterLondonAuthority/dfl-ckanext/pull/2#pullrequestreview-1466076912) for more information

madelinekosse committed:
DAT-344 - log correct index of search result

- Don't add one to index as it already starts at 1
- Use number of results per page from config instead of hard-coding

Co-authored-by: Kiran Joshi <kiran.j88@gmail.com>

madelinekosse committed:
DAT-344 - remove unused module tag

Co-authored-by: Kiran Joshi <kiran.j88@gmail.com>

kjoshi reviewed (approved):
Looks good to me! 👍
